### ISSUE TITLE
Arrow navigation issue with single-character text node adjacent to inline element

### ISSUE THREAD

**Issue by dmarkow** [ianstormtaylor/slate#3458]
#### Do you want to request a _feature_ or report a _bug_?

Bug

#### What's the current behavior?

https://codesandbox.io/s/slate-reproductions-qdi9i

Click anywhere in the first link. Use the right/left arrow keys to move around. You cannot move the cursor to the left of the first x or the right of the last x. Clicking before/after them works fine though.

If the text nodes have more than 1 character, they work fine (see the second editor in the example)

Slate: 0.57.1
Browser: Chrome
OS: Mac

#### What's the expected behavior?

Arrow keys should navigate to the end of the line

**Comment by dmarkow** [comment]
Actually, it seems to affect any adjacent nodes, not just inline. So for example, have A**B**C where the B is bolded, doesn't allow arrow navigation before the A or after the C.

**Comment by davetodd** [comment]
I can also reproduce this on the rich text example at https://www.slatejs.org/examples/richtext.

I've tracked this down as far as `Editor.positions()`. That function doesn't return the `{ offset: 0, path: [0,0] }` position if the first node is only a single character. It's pretty easy to create a failing unit test for this, but it's not obvious to me how to fix it. 

As far as I can tell, the previous node is "consuming" the first/last node's text, such that this line causes it to short-circuit before returning the final position: https://github.com/ianstormtaylor/slate/blob/master/packages/slate/src/interfaces/editor.ts#L1065

**Comment by majelbstoat** [comment]
It is also the case that if you take the steps in the Rich Text example to :

* start in an empty paragraph
* type ABC
* bold B
* move cursor between B and C
* press Enter

the paragraph splits, but the cursor stays on the same line. Feels like the same root cause?

**Comment by beorn** [comment]
I also discovered this bug, I think, and made a different test scenario with debug info: https://codesandbox.io/s/slate-reproductions-wbub5

**Comment by beorn** [comment]
> I can also reproduce this on the rich text example at https://www.slatejs.org/examples/richtext.
> 
> I've tracked this down as far as `Editor.positions()`. That function doesn't return the `{ offset: 0, path: [0,0] }` position if the first node is only a single character. It's pretty easy to create a failing unit test for this, but it's not obvious to me how to fix it.
> 
> As far as I can tell, the previous node is "consuming" the first/last node's text, such that this line causes it to short-circuit before returning the final position: https://github.com/ianstormtaylor/slate/blob/master/packages/slate/src/interfaces/editor.ts#L1065

The PR above (#3644) does fix this specific case, but I don't think the other arrow navigation issues are related as they behave differently in different browsers.  See my test scenario at https://codesandbox.io/s/slate-reproductions-wbub5

**Comment by beorn** [comment]
Just updated the PR (#3644) - I think all of the above bugs might be fixed by this now.

**Comment by beorn** [comment]
I tracked down another problem: Moving the cursor around results in a `Transforms.move()` with `unit='character'`, which if the neighboring nodes doesn't have any characters will not result in any move.  Is there any reason why we don't default to `unit='offset'` for this and other keyboard input stuff (like `delete*` transforms)?

**Comment by BrentFarese** [comment]
For number (3), perhaps we can make the default behavior the way it is now and simply include an option to disable that behavior if the Slate user wants to?  In our implementation, and in many modern editors, the `Editor.insertText` behavior for inlines is correct and you can't type at the end of the inline but I have seen use cases where Slate users don't want that behavior.

**Comment by beorn** [comment]
@BrentFarese Fortunately, it is relatively easy to override `Editor.insertText` in your plugin to create 'sticky inlines', but you do end up duplicating half of the logic that's in the default `Editor.insertText`, which is not ideal (see example below).  While (1) and (2) are not easy to fix without updating SlateJS code, and without those, cursor movement and editing of inline elements is still problematic - maybe even more so due to inconsistencies caused by (2).

```js
export function withCoreOverrides(editor) {
  const { insertText } = editor

  editor.insertText = (text, options = {}) => {
    const { selection, marks } = editor
    let { at = selection } = options
    if (Range.isRange(at) && !Range.isCollapsed(at)) {
      return insertText(text, options)
    }
    at = at.anchor
    if (marks) {
      const node = { text, ...marks }
      Transforms.insertNodes(editor, node)
    } else {
      const { path, offset } = at
      editor.apply({ type: 'insert_text', path, offset, text })
    }
    editor.marks = null
  }
  return editor
}
```

**Comment by BrentFarese** [comment]
Yes that's true people can write their own plugin of `insertText` (a plugin can be written the other way too so you can type at the end of an inline) but the change in (3) would be a breaking change.  I would suggest we do something more like `insertText: (text: string, enableInlineTyping?: boolean)` where the `enableInlineTyping` flag can be used within `insertText` to allow typing at the end of an inline.  The `enableInlineTyping` flag can be defaulted to `false` so it's optional and has no impact on those who rely on this feature already.

**Comment by beorn** [comment]
You are right!  But this in fact means that none of the other logic built on top of `*positions` handle `unit=line` correctly, as they all rely on it; `Editor.before`, `Editor.delete`, etc.

I think it would be an important thing to implement, but I assume it'll require rendering/DOM knowledge?

**Comment by davetodd** [comment]
> The PR above (#3644) does fix the case where a position might be missed at the end. Can you supply a test case where the position is missed at the beginning?

@beorn The test you have in your PR is actually the one I was thinking of, though if you wanted to be extra complete, you could swap a + b so that the inline is the first element. Both cases worked when I tested your branch locally. The bug + your fix also applied to a single character adjacent to another leaf, rather than an inline, which is 👍 

I did notice while testing this that if you insert a break before a new leaf, the cursor remains on the original line. But if you insert a break before an inline, the cursor moves to the new line. Here's a fork of your sandbox that shows that: https://codesandbox.io/s/slate-reproductions-dfgd8 ...it looks like the positions array is the same in both cases, so this seems pretty unrelated and more likely related to Transforms.splitNodes().
